src/audit/unpinned_uses.rs

```diff
@@ -21,15 +23,31 @@ impl WorkflowAudit for UnpinnedUses {
             return Ok(vec![]);
         };
 
-        if uses.unpinned() {
+        // Check twice, complain once
+        let mut found_unpinned = false;
+        let mut annotation = "";
+        if self.state.pedantic {
+            if uses.unpinned_sha() {
+                found_unpinned = true;
+                annotation = "action is not pinned to a hash ref";
+            }
+        } else {
+            if uses.unpinned() {
+                found_unpinned = true;
+                annotation = "action is not pinned to a tag, branch, or hash ref";
+            }
+        }
+
+        // If one of the checks failed, add a finding
+        if found_unpinned {
```

Comment:
FWIW, I think the idiomatic Rust structure for this would be something like:

```rust
let annotation = if { ... } else if { ... } else { continue; };
```

which would avoid the need for the `found_unpinned` variable and the `mut`s. I can send a commit to this branch showing what I mean If I'm not explaining that well :slightly_smiling_face:

Reply:
Actually, thinking about this a bit more -- perhaps it makes sense to emit 2 separate findings here, one for the non-pedantic case and one for the pedantic case? My thinking is that `--pedantic` shouldn't change the behavior shape of findings, it should only be strictly additive to the findings set.

Thoughts? I think that would also simplify the code slightly, since there's be no need to parametrize the annotation.

Reply:
Hm. So in the pedantic case, `uses: some/action` would emit two findings? I don't think that's particularly useful as a user, but I might be missing a use case for it.

Maybe this is just an abuse of `--pedantic` because we don't have a `--strict` or the like. In this specific case, we're not adding an audit, really, just intensifying one that already exists.

Having typed that out, my weak preference is to keep it as-is, with the more idiomatic style you suggested.

Reply:
> Maybe this is just an abuse of `--pedantic` because we don't have a `--strict` or the like. In this specific case, we're not adding an audit, really, just intensifying one that already exists.

That's a good point. I don't have a super strong opinion or intuition here, so we can keep it as one finding for now and see what happens as I tweak the pedantic/sensitivity model 🙂

Reply:
Oh, here's another thought: could we use different severities for the two so the user can use `--min-severity` to decide how strict they want to be? This is essentially the sensitivity model discussion, IMO

So I'd put the sha check at informative or low, and the entirely unpinned at medium or high. My inclination is to err on the stricter side, since this is an easy avenue for supply chain attacks, and go with unhashed is low and unpinned is high, but I don't have a strong take on that.